test/integration_test.go

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package test
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/pborman/uuid"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/cadence"
+	"go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/client"
+	"go.uber.org/cadence/worker"
+	"go.uber.org/cadence/workflow"
+	"go.uber.org/goleak"
+	"go.uber.org/zap"
+)
+
+type IntegrationTestSuite struct {
+	*require.Assertions
+	suite.Suite
+	config       Config
+	rpcClient    *rpcClient
+	libClient    client.Client
+	activities   *Activities
+	workflows    *Workflows
+	worker       worker.Worker
+	seq          int64
+	taskListName string
+}
+
+const (
+	ctxTimeout                 = 10 * time.Second
+	domainName                 = "integration-test-domain"
+	domainCacheRefreshInterval = 11 * time.Second
+)
+
+func TestIntegrationSuite(t *testing.T) {
+	suite.Run(t, new(IntegrationTestSuite))
+}
+
+func (ts *IntegrationTestSuite) SetupSuite() {
+	ts.Assertions = require.New(ts.T())
+	ts.config = newConfig()
+	ts.activities = &Activities{}
+	ts.workflows = &Workflows{}
+	ts.register()
+}
+
+func (ts *IntegrationTestSuite) TearDownSuite() {
+	// sleep for a while to allow the pollers to shutdown
+	// then assert that there are no lingering go routines
+	time.Sleep(time.Minute + time.Second)
```

Comment:
yes, there is an api to stop pollers and when worker.Stop() is called, that api will be called. Problem is - if a goroutine is blocked on a call to PollForDecisionTask() - it won't return from that call when poller.Stop() is called. This is because, we use shutdown channels and don't actually cancel the context that the poller is blocked on. Having shutdown channels does make sense from a simplicity perspective but it just means, we will shutdown only after the blocking call returns.

Reply:
Could we close the context on shutdown channel close?

Reply:
its complicated :) A somewhat more detailed explanation:

TaskPoller interface:
https://github.com/uber-go/cadence-client/blob/master/internal/internal_task_pollers.go#L52
Notice, it doesn't have Stop() method on it. Also notice, it has ProcessTask() in it. Poll() will call PollForActivityTask or PollForDecisionTask. ProcessTask will call the corresponding activity or dispatcher and subsequently call RespondActivityTask() or RespondDecisionTask().

Go Routines:
https://github.com/uber-go/cadence-client/blob/master/internal/internal_worker_base.go#L223

baseWorker creates a runPoller()  goroutine for every poller. This routine in turn spins up a goroutine to call PollTask(). 
baseWorker also has a goroutine for task processing which calls ProcessTask()
https://github.com/uber-go/cadence-client/blob/master/internal/internal_worker_base.go#L258

Fix:
The fix involves a bit of refactoring since poller does both polling / processing and can block on the following:
 - PollForActivity, PollForDecision
 - RespondDecision, RespondActivity 
 - ActivityHandler itself (outside of our control)

I punted on this initially because this deserves its own separate PR. But now, I made this change. See if that makes sense. Now everything shuts down in 10s or so.

Reply:
Wow! Nice!